db/db_wal_test.cc

```diff
@@ -1565,6 +1561,65 @@ TEST_P(DBWALTestWithParams, kSkipAnyCorruptedRecords) {
   }
 }
 
+class DBWALTestWithParamsRecycling
+    : public DBWALTestBase,
+      public ::testing::WithParamInterface<
+          std::tuple<bool, int, WALRecoveryMode>> {
+ public:
+  DBWALTestWithParamsRecycling()
+      : DBWALTestBase("/db_wal_test_with_params_recovery_hole") {}
+};
+
+INSTANTIATE_TEST_CASE_P(
+    Wal, DBWALTestWithParamsRecycling,
+    ::testing::Combine(
+        ::testing::Bool(), ::testing::Values(0, 2),
+        ::testing::Values(WALRecoveryMode::kTolerateCorruptedTailRecords,
+                          WALRecoveryMode::kAbsoluteConsistency,
+                          WALRecoveryMode::kPointInTimeRecovery,
+                          WALRecoveryMode::kSkipAnyCorruptedRecords)));
+
+TEST_P(DBWALTestWithParamsRecycling, RecoveryHole) {
+  Options options = CurrentOptions();
+  options.track_and_verify_wals_in_manifest = std::get<0>(GetParam());
+  options.recycle_log_file_num = std::get<1>(GetParam());
+  options.wal_recovery_mode = std::get<2>(GetParam());
+
+  CreateAndReopenWithCF({"pikachu"}, options);
+  ASSERT_OK(Put(1, "ok", "ok"));
+
+  // Record the offset at this point
+  Env* env = options.env;
+  uint64_t wal_file_id = dbfull()->TEST_LogfileNumber();
+  std::string fname = LogFileName(dbname_, wal_file_id);
+  uint64_t offset_to_corrupt;
+  ASSERT_OK(env->GetFileSize(fname, &offset_to_corrupt));
+  ASSERT_GT(offset_to_corrupt, 0);
+  ASSERT_OK(Put(0, "key1", "wal1"));
+  ASSERT_OK(Flush(1));
+  ASSERT_OK(Put(0, "key2", "wal2"));
+
+  // Corrupt WAL1 at location of key1
+  ASSERT_OK(test::CorruptFile(env, fname, static_cast<int>(offset_to_corrupt),
```

Comment:
This is probably corrupting the WAL fragment header for key1. What if we corrupt one byte of the value? For example take the file length after putting "key1" -> "wal1" and corrupt the final byte. I think there's no WAL fragment trailer.

Reply:
I updated the unit test to corrupt the end byte and a middle byte in addition to the first byte of the record. Still, it does not seem to permit a hole in recovery.